engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,832 @@
+//   │    │                       │                              │                                │
+//   │    │ Redirect              │                              │                                │
+//   │    │                       │                              │                                │
+//   │◄───┘                       │                              │                                │
+//   │                            │                              │                                │
+
+type tableType int
+
+const (
+	normal tableType = iota
+	conflict
+	final
+)
+
+// TableAgent defines an interface for checkpoint.
+type TableAgent interface {
+	FetchAllDoTables(ctx context.Context, cfg *config.JobCfg) (map[metadata.TargetTable][]metadata.SourceTable, error)
+	FetchTableStmt(ctx context.Context, jobID string, cfg *config.JobCfg, sourceTable metadata.SourceTable) (string, error)
+}
+
+// DDLCoordinator is a coordinator for ddl.
+type DDLCoordinator struct {
+	mu          sync.Mutex
+	tables      map[metadata.TargetTable]map[metadata.SourceTable]struct{}
+	shardGroups map[metadata.TargetTable]*shardGroup
+	pendings    sync.WaitGroup
+	logger      *zap.Logger
+
+	kvClient   metaModel.KVClient
+	tableAgent TableAgent
+	jobID      string
+	jobStore   *metadata.JobStore
+}
+
+// NewDDLCoordinator creates a new DDLCoordinator.
+func NewDDLCoordinator(jobID string, kvClient metaModel.KVClient, tableAgent TableAgent, jobStore *metadata.JobStore, pLogger *zap.Logger) *DDLCoordinator {
+	return &DDLCoordinator{
+		tableAgent:  tableAgent,
+		tables:      make(map[metadata.TargetTable]map[metadata.SourceTable]struct{}),
+		jobID:       jobID,
+		kvClient:    kvClient,
+		shardGroups: make(map[metadata.TargetTable]*shardGroup),
+		jobStore:    jobStore,
+		logger:      pLogger.With(zap.String("component", "ddl_coordinator")),
+	}
+}
+
+// Clear clears metadata.
+func (c *DDLCoordinator) Clear(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	return metadata.DelAllDroppedColumns(ctx, c.kvClient)
+}
+
+// Reset resets the shard group.
+func (c *DDLCoordinator) Reset(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// wait all pending ddls coordinated.
+	c.pendings.Wait()
+
+	jobCfg, err := c.jobStore.GetJobCfg(ctx)
+	if err != nil {
+		return err
+	}
+	// fetch all tables which need to be coordinated.
+	tables, err := c.tableAgent.FetchAllDoTables(ctx, jobCfg)
+	if err != nil {
+		return err
+	}
+
+	for targetTable, sourceTables := range tables {
+		c.tables[targetTable] = make(map[metadata.SourceTable]struct{}, 0)
+		for _, sourceTable := range sourceTables {
+			c.tables[targetTable][sourceTable] = struct{}{}
+		}
+	}
+	return nil
+}
+
+// Coordinate coordinates ddls.
+func (c *DDLCoordinator) Coordinate(ctx context.Context, item *metadata.DDLItem) ([]string, optimism.ConflictStage, error) {
+	c.mu.Lock()
+	c.pendings.Add(1)
```

Comment:
the problem is the race accessing of c.pendings. Even if we use lock to avoid simultaneously accessing, will it happen that `Coordinate` is called after `Clear`?

Reply:
Clear only can be called when cancel job, in that time all the worker should be closed, or the jobmaster will be forcely close